src/components/change-view/change-view.jsx

```diff
@@ -43,55 +44,43 @@ const diffSettingsStorage = 'edgi.wm.ui.diff_settings';
  * @param {ChangeViewProps} props
  */
 export default class ChangeView extends React.Component {
+  // TODO: This fires more often than is probably necessary.
+  // see about alternatives to getDerivedStateFromProps
+  // https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html
```

Comment:
OK, having read this blog post for the umpteenth time, I’m pretty sure what we have here is actually a textbook case of exactly when we *should* use `getDerivedStateFromProps()`. (I’m sure we have some bad examples in the codebase, too, just not this one.)

What we’re fundamentally doing here is validating chosen `diffType`. From version to version (i.e. when `props.from` or `props.to` change), the media types may change and the previously selected diff may become invalid, at which point we have to calculate a new `state.diffType` value based on the props. Pretty consistent with that article and the React docs:

> `getDerivedStateFromProps` exists for only one purpose. It enables a component to update its internal state as the result of **changes in props.**

Anyway, the previous version of this function also went out of its way to be careful about the potential issues by only returning new state information if `state.diffType` was invalid. That’s no longer the case in the new function — a more equivalent version would be:

```js
static getDerivedStateFromProps (props, state) {
  if (props.from && props.to) {
    const diffType = findDiffType(props.from, props.to, props.page, state.diffType);
    if (diffType !== state.diffType) return { diffType };
  }

  return null;
}
```

**Buuuuut** I’m not sure it’s a big deal, since the worst that can happen in the current case is that we tell React to update `state.diffType` to what it already is (since `findDiffType` will always return the 4th argument unless it was invalid or not set).

Reply:
NB: I guess I could have added some comments that explained the above bit about how and when this code is relevant (validating the choice of diff when props change). Maybe a good idea to do here, assuming what I wrote above actually makes sense to you.

Reply:
Also: renaming `findDiffType` → `ensureValidDiffType` might help out here, too, since what that function is really doing is returning the requested diff type if it’s valid, or returning the best possible valid type otherwise.

Reply:
Oh ok I think I more fully understand now. I've added that conditional check back in and renamed the helper function. I think the cleaner solution in this instance might be lifting state or pulling out state, which would probably fall under this issue: https://github.com/edgi-govdata-archiving/web-monitoring-ui/issues/126

Reply:
Yeah, I suppose it’s definitely a little inconsistent that `from`/`to` are controlled properties, but `diffType` is uncontrolled.